### ISSUE TITLE
`Index.sum_over` into few keys is ~5000x slower in float32 on GPU (jaxlib 0.10.2+ f32 scatter-add CAS regression)

### ISSUE THREAD

**Issue by jonkhler** [cusp-ai-oss/kUPS#248]
## Summary

`Index.sum_over` lowers to `jax.ops.segment_sum`. On XLA:GPU with jaxlib 0.10.2 to 0.11.0, a float32 segment sum of many values into few segments is up to ~5000x slower than the same reduction into a few thousand segments. The segment count drives the penalty, not the value count. float64 is not affected. x64 mode hides the problem.

## Impact

This is the shape of every "reduce per-edge value to per-system value" step. `lennard_jones_energy` (`src/kups/potential/classical/lennard_jones.py:172`) reduces ~400k edge energies into 1–24 systems:

```python
total_energies = graph.edge_batch_mask.sum_over(edge_energy) / 2
```

A batched LJ force+energy step at 24 systems x 256 atoms in float32 costs ~900 ms. This one reduction causes almost all of it. `observables/pressure.py` and `observables/stress.py` reduce with the same shape.

## Environment

- kups 1.0.3 (`1b3d70d`)
- jax / jaxlib 0.11.0, CUDA 13
- NVIDIA L4, driver 580.173.02
- Python 3.12

## Steps to reproduce

Run this script on a CUDA GPU (no kups needed):

```python
import jax, jax.numpy as jnp, numpy as np, time

def bench(f, *a, n=8):
    o = f(*a); jax.block_until_ready(o); t0 = time.time()
    for _ in range(n): o = f(*a)
    jax.block_until_ready(o); return (time.time() - t0) / n * 1e3

E, FEW, MANY = 393_216, 24, 6_144          # 24 systems x 256 particles

vals = jnp.asarray(np.random.randn(E), dtype=jnp.float32)
few  = jnp.asarray(np.repeat(np.arange(FEW), E // FEW), dtype=jnp.int32)
many = jnp.asarray(np.random.randint(0, MANY, E), dtype=jnp.int32)

to_few  = jax.jit(lambda v, i: jax.ops.segment_sum(v, i, num_segments=FEW))
to_many = jax.jit(lambda v, i: jax.ops.segment_sum(v, i, num_segments=MANY))
staged  = jax.jit(lambda v, i: jax.ops.segment_sum(v, i, num_segments=MANY)
                                 .reshape(FEW, -1).sum(-1))

print(f"f32 -> {FEW:5d} segments : {bench(to_few,  vals, few ):9.3f} ms")
print(f"f32 -> {MANY:5d} segments : {bench(to_many, vals, many):9.3f} ms")
print(f"f32 two-stage           : {bench(staged,  vals, many):9.3f} ms")
```

## Observed result

| reduction | float32 |
|---|---|
| 393,216 -> 24 segments | **905.143 ms** |
| 393,216 -> 6,144 segments | 0.181 ms |
| two-stage (-> 6,144, then `reshape(24,-1).sum(-1)`) | 0.180 ms |

## Root cause (upstream)

The regression is in the LLVM NVPTX backend, pinned into XLA and jaxlib 0.10.2+. LLVM stopped the lowering of f32 `atomicrmw fadd` to the native atomic in IEEE denormal mode. XLA compiles in IEEE mode by default. The scatter-add kernel then contains a compare-and-swap retry loop. Colliding threads retry. Few segments cause maximal collision.

PTX evidence from this machine, same HLO for both dtypes:

- float32: `atom.relaxed.sys.global.cas.b32` (CAS retry loop, slow)
- float64: `atom.relaxed.sys.global.add.f64` (native atomic, fast)

Tracked upstream:

- jax-ml/jax#38806 (open; full root cause and workaround)
- openxla/xla#46225 (closed as duplicate of the above)

LLVM reverted the default in llvm/llvm-project#206154 (2026-07-17). The revert reaches jaxlib with a future LLVM integrate. Do not file a new upstream issue.

Workaround, verified on this machine (905 ms -> 0.651 ms):

```
XLA_FLAGS="--xla_backend_extra_options=-nvptx-allow-ftz-atomics"
```

## Possible fix

Stage the reduction inside `Index.sum_over`. Spread each label over `sub` interleaved sub-buckets. Reduce into `num_labels * sub >= 4096` intermediate segments. Then reduce the partials densely. This is correct for ragged systems. No API change. All callers benefit.

```diff
diff --git a/src/kups/core/data/index.py b/src/kups/core/data/index.py
index 694a16b..a607c44 100644
--- a/src/kups/core/data/index.py
+++ b/src/kups/core/data/index.py
@@ -41,6 +41,16 @@ type PyTree = Any
 if TYPE_CHECKING:
     from kups.core.data.table import Table
 
+_STAGED_SEGMENT_SUM_TARGET = 4096
+"""Minimum segment count below which ``Index.sum_over`` stages its reduction.
+
+XLA:GPU lowers a segment sum to scatter-add, which serializes colliding
+updates; summing many values into few segments is ~1000x slower in float32
+than the same reduction into a few thousand segments. ``sum_over`` therefore
+spreads each label over enough interleaved sub-buckets to reach roughly this
+many intermediate segments and then reduces the partials densely.
+"""
+
 
 class SupportsDunderLT(Protocol):
     def __lt__(self, other: Any, /) -> bool: ...
@@ -520,6 +530,10 @@ class Index[Key: SupportsSorting]:
     def sum_over(self, array: Array) -> Table[Key, Array]:
         """Sums ``array`` values grouped by this index via segment sum.
 
+        When many values reduce into few keys, the reduction is staged through
+        ``_STAGED_SEGMENT_SUM_TARGET`` interleaved sub-buckets to avoid the
+        XLA:GPU scatter-add serialization on colliding segment ids.
+
         Args:
             array: Array with leading dimension matching ``self.indices``.
 
@@ -528,11 +542,19 @@ class Index[Key: SupportsSorting]:
         """
         from kups.core.data.table import Table
 
-        return Table(
-            self.keys,
-            jax.ops.segment_sum(array, self.indices, self.num_labels, mode="drop"),
-            _cls=self._cls,
-        )
+        num_labels = self.num_labels
+        sub = -(-_STAGED_SEGMENT_SUM_TARGET // num_labels) if num_labels > 0 else 1
+        if self.indices.ndim != 1 or sub <= 1 or self.indices.size <= num_labels * sub:
+            summed = jax.ops.segment_sum(array, self.indices, num_labels, mode="drop")
+        else:
+            # Value ``r`` with label ``i`` lands in sub-bucket ``i*sub + r%sub``;
+            # OOB labels (>= num_labels, or negative) stay OOB and are dropped.
+            lanes = jnp.arange(self.indices.size, dtype=self.indices.dtype) % sub
+            partials = jax.ops.segment_sum(
+                array, self.indices * sub + lanes, num_labels * sub, mode="drop"
+            )
+            summed = partials.reshape(num_labels, sub, *partials.shape[1:]).sum(axis=1)
+        return Table(self.keys, summed, _cls=self._cls)
 
     def max_over(self, array: Array) -> Table[Key, Array]:
         """Takes max of ``array`` values grouped by this index via segment max.
```

Keep the fix after the jaxlib regression is gone. The staged form (0.385 ms) is faster than the contended native atomic (0.651 ms with the workaround flag).

## Verification

Measured through `Index.sum_over` on the L4, 393,216 float32 values into 24 keys:

| path | time |
|---|---|
| direct (current `main`) | 904.917 ms |
| staged (this fix) | **0.385 ms** |

Correctness checks against the direct path, all pass:

- ragged label counts and value counts, with trailing dimensions
- out-of-bounds ids (`== num_labels` sentinel, negative): dropped in both paths
- int32 values: bit-exact match
- under `jit` and `grad`
- 2-D `indices`: keeps the direct path

**Comment by n-gao** [comment]
@jonkhler I think this should be a JAX issue and not a kUPS issue, I would avoid fine tuning the code to specific JAX versions and their bugs.

**Comment by n-gao** [comment]
Fixed in #272  and jax 0.11.1
